aten/src/ATen/ParallelOpenMP.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+#include <ATen/ATen.h>
+
+#include <cstddef>
+#include <exception>
+
+#ifdef _OPENMP
+#define INTRA_OP_PARALLEL
+
+#include <omp.h>
+#endif
+
+namespace at {
+
+template <class F>
+void parallel_for(
+    const int64_t begin,
+    const int64_t end,
+    const int64_t grain_size,
+    const F& f) {
+#ifdef _OPENMP
+  std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
+  std::exception_ptr eptr;
+#pragma omp parallel if (!omp_in_parallel() && ((end - begin) >= grain_size))
+  {
+    int64_t num_threads = omp_get_num_threads();
+    int64_t tid = omp_get_thread_num();
+    int64_t chunk_size = divup((end - begin), num_threads);
+    int64_t begin_tid = begin + tid * chunk_size;
+    if (begin_tid < end) {
+      try {
```

Comment:
Is a try catch block necessary here? Isn't it quite expensive?

Reply:
could be expensive, in this PR I just refactored code into separate files though, to allow adding of other backends

Reply:
Ok, we should mark this for an immediate follow up PR, since it could slow down things a lot.